File: frontend/lib/models/music.dart
enum InputType { voice, text, image }

enum MusicStatus { generating, completed, failed }

class Music {
  final int id;
  final int userId;
  final String title;
  final String? description;
  final InputType inputType;
  final String? inputContent;     // 文字内容或语音转文字
  final String? inputImageUrl;    // 图片URL（如果是图片输入）
  final List<String>? emotionTags;
  final String? coverUrl;         // 封面图片URL
  final String? aiAnalysis;
  final String musicUrl;
  final String musicFormat;
  final int duration;
  final int fileSize;
  final int bpm;
  final String? genre;
  final List<String>? instruments;
  final MusicStatus status;
  final bool isPublic;
  final bool isFavorite;          // 是否收藏
  final int playCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  Music({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    required this.inputType,
    this.inputContent,
    this.inputImageUrl,
    this.emotionTags,
    this.coverUrl,
    this.aiAnalysis,
    required this.musicUrl,
    required this.musicFormat,
    required this.duration,
    required this.fileSize,
    required this.bpm,
    this.genre,
    this.instruments,
    required this.status,
    required this.isPublic,
    this.isFavorite = false,
    required this.playCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Music.fromJson(Map<String, dynamic> json) {
    return Music(
      id: json['id'],
      userId: json['user_id'],
      title: json['title'],
      description: json['description'],
      inputType: _parseInputType(json['input_type']),
      inputContent: json['input_content'],
      inputImageUrl: json['input_image_url'],
      emotionTags: json['emotion_tags'] != null
          ? List<String>.from(json['emotion_tags'])
          : null,
      coverUrl: json['cover_url'],
      aiAnalysis: json['ai_analysis'],
      musicUrl: json['music_url'] ?? '',
      musicFormat: json['music_format'] ?? 'mp3',
      duration: json['duration'] ?? 0,
      fileSize: json['file_size'] ?? 0,
      bpm: json['bpm'] ?? 120,
      genre: json['genre'],
      instruments: json['instruments'] != null
          ? List<String>.from(json['instruments'])
          : null,
      status: _parseStatus(json['status']),
      isPublic: json['is_public'] ?? false,
      isFavorite: json['is_favorite'] ?? false,
      playCount: json['play_count'] ?? 0,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  static InputType _parseInputType(String? type) {
    switch (type) {
      case 'voice':
        return InputType.voice;
      case 'text':
        return InputType.text;
      case 'image':
        return InputType.image;
      default:
        return InputType.text;
    }
  }

  static MusicStatus _parseStatus(String? status) {
    switch (status) {
      case 'generating':
        return MusicStatus.generating;
      case 'completed':
        return MusicStatus.completed;
      case 'failed':
        return MusicStatus.failed;
      default:
        return MusicStatus.generating;
    }
  }

  String get inputTypeString {
    switch (inputType) {
      case InputType.voice:
        return 'voice';
      case InputType.text:
        return 'text';
      case InputType.image:
        return 'image';
    }
  }

  String get inputTypeIcon {
    switch (inputType) {
      case InputType.voice:
        return '🎤';
      case InputType.text:
        return '✍️';
      case InputType.image:
        return '🖼️';
    }
  }

  String get inputTypeLabel {
    switch (inputType) {
      case InputType.voice:
        return '语音';
      case InputType.text:
        return '文字';
      case InputType.image:
        return '图片';
    }
  }

  String get durationString {
    final minutes = duration ~/ 60;
    final seconds = duration % 60;
    return '${minutes.toString().padLeft(2, '0')}:${seconds.toString().padLeft(2, '0')}';
  }

  /// 从 emotionTags 获取主要情绪
  String get primaryEmotion {
    if (emotionTags != null && emotionTags!.isNotEmpty) {
      return emotionTags!.first;
    }
    return 'calm';
  }

  Music copyWith({
    int? id,
    int? userId,
    String? title,
    String? description,
    InputType? inputType,
    String? inputContent,
    String? inputImageUrl,
    List<String>? emotionTags,
    String? coverUrl,
    String? aiAnalysis,
    String? musicUrl,
    String? musicFormat,
    int? duration,
    int? fileSize,
    int? bpm,
    String? genre,
    List<String>? instruments,
    MusicStatus? status,
    bool? isPublic,
    bool? isFavorite,
    int? playCount,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Music(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      title: title ?? this.title,
      description: description ?? this.description,
      inputType: inputType ?? this.inputType,
      inputContent: inputContent ?? this.inputContent,
      inputImageUrl: inputImageUrl ?? this.inputImageUrl,
      emotionTags: emotionTags ?? this.emotionTags,
      coverUrl: coverUrl ?? this.coverUrl,
      aiAnalysis: aiAnalysis ?? this.aiAnalysis,
      musicUrl: musicUrl ?? this.musicUrl,
      musicFormat: musicFormat ?? this.musicFormat,
      duration: duration ?? this.duration,
      fileSize: fileSize ?? this.fileSize,
      bpm: bpm ?? this.bpm,
      genre: genre ?? this.genre,
      instruments: instruments ?? this.instruments,
      status: status ?? this.status,
      isPublic: isPublic ?? this.isPublic,
      isFavorite: isFavorite ?? this.isFavorite,
      playCount: playCount ?? this.playCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: backend/app/models/music.py
"""
音乐相关数据模型 - 完整修复版
"""
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text, Enum, ForeignKey, JSON, UniqueConstraint
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.database import Base
import enum


class InputType(str, enum.Enum):
    voice = "voice"
    text = "text"
    image = "image"


class MusicStatus(str, enum.Enum):
    generating = "generating"
    completed = "completed"
    failed = "failed"


class EmotionType(str, enum.Enum):
    happy = "happy"
    calm = "calm"
    sad = "sad"
    energetic = "energetic"
    nostalgic = "nostalgic"


class Music(Base):
    __tablename__ = "musics"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text)
    
    # 输入信息
    input_type = Column(Enum(InputType), nullable=False)
    input_content = Column(Text)
    emotion_tags = Column(JSON)
    primary_emotion = Column(String(50), index=True)
    ai_analysis = Column(Text)
    
    # 音乐文件信息
    music_url = Column(String(500), nullable=False)
    cover_url = Column(String(500))
    music_format = Column(String(10), default="mp3")
    duration = Column(Integer, default=0)
    file_size = Column(Integer, default=0)
    
    # 音乐参数
    bpm = Column(Integer, default=120)
    genre = Column(String(100))
    instruments = Column(JSON)
    
    # 状态
    status = Column(Enum(MusicStatus), default=MusicStatus.generating, index=True)
    is_public = Column(Boolean, default=False)
    play_count = Column(Integer, default=0)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # 关系
    user = relationship("User", back_populates="musics")
    collections = relationship("Collection", back_populates="music", cascade="all, delete-orphan")
    favorites = relationship("Favorite", back_populates="music", cascade="all, delete-orphan")


class Collection(Base):
    """用户收藏夹/文件夹"""
    __tablename__ = "collections"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    music_id = Column(Integer, ForeignKey("musics.id", ondelete="CASCADE"), nullable=False)
    folder_name = Column(String(100), default="default")
    note = Column(Text)
    tags = Column(JSON)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    user = relationship("User", back_populates="collections")
    music = relationship("Music", back_populates="collections")


class Favorite(Base):
    """用户喜欢/收藏的音乐（简单收藏，不分文件夹）"""
    __tablename__ = "favorites"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    music_id = Column(Integer, ForeignKey("musics.id", ondelete="CASCADE"), nullable=False, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 确保用户不能重复收藏同一首歌
    __table_args__ = (
        UniqueConstraint('user_id', 'music_id', name='unique_user_music_favorite'),
    )
    
    # 关系
    user = relationship("User", back_populates="favorites")
    music = relationship("Music", back_populates="favorites")


class Playlist(Base):
    __tablename__ = "playlists"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    name = Column(String(255), nullable=False)
    description = Column(Text)
    cover_url = Column(String(500))
    is_public = Column(Boolean, default=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # 关系
    user = relationship("User", back_populates="playlists")
    items = relationship("PlaylistItem", back_populates="playlist", cascade="all, delete-orphan")


class PlaylistItem(Base):
    __tablename__ = "playlist_items"
    
    id = Column(Integer, primary_key=True, index=True)
    playlist_id = Column(Integer, ForeignKey("playlists.id", ondelete="CASCADE"), nullable=False)
    music_id = Column(Integer, ForeignKey("musics.id", ondelete="CASCADE"), nullable=False)
    position = Column(Integer, default=0)
    added_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    playlist = relationship("Playlist", back_populates="items")
    music = relationship("Music")


class GenerationLog(Base):
    __tablename__ = "generation_logs"
    
    id = Column(Integer, primary_key=True, index=True)
    music_id = Column(Integer, ForeignKey("musics.id", ondelete="CASCADE"), nullable=False)
    step = Column(String(100))
    status = Column(String(50))
    message = Column(Text)
    progress = Column(Integer, default=0)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    music = relationship("Music")
